### DISCUSSION TITLE
[Auto Versioning] Updating JSONB field content to bump model version

### DISCUSSION THREAD

**Discussion by funkindy** [sqlalchemy/sqlalchemy#11276] [Usage Questions] [answered]
### Describe the use case

I am using auto versioning feature of sqlalchemy:

```
class Counterparty(Base):
    id: ...
    legal_name: Mapped[str] = mapped_column(sa.String(255), nullable=False)
    basic_data: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=False)
    version = mapped_column(Integer, nullable=False)

    __mapper_args__ = {'version_id_col': version}
```

If i update legal_name, vesion get bumped and it is okay. This test passes:

```
async def test_version_bumps_on_update(counterparty, session):
    counterparty.legal_name = 'New Legal Name'
    await session.commit()
    assert counterparty.version == 2
```

But if i update the content of JSONB, the version is still == 1 and test fails:

```
async def test_version_bumps_on_update(counterparty, session):
    counterparty.basic_data['address'] = 'New Address'
    await session.commit()
    assert counterparty.version == 2
```
error:

```
>       assert counterparty.version == 2
E       assert 1 == 2
E        +  where 1 = <models.counterparty.Counterparty object at 0x106fe7ec0>.version
```

I`d like the auto versioning feature to also bumps version on JSONB internal updates, if it is possible.

### Databases / Backends / Drivers targeted

PG 16, asyncpg, python 3.12

### Example Use

Please take a look at the faulty example in description

**Comment by zzzeek** [comment, marked as answer]
this operation:

    counterparty.basic_data['address'] = 'New Address'

is not picked up by the unit of work, unless you use the [in place, mutation-tracking extension](https://docs.sqlalchemy.org/en/20/orm/extensions/mutable.html) for your JSONB column.

**Comment by funkindy** [comment]
Wow, great!
Solved just with one line change:

```
    basic_data: Mapped[MutableDict[str, Any]] = mapped_column(
        MutableDict.as_mutable(JSONB), nullable=False, doc='Основные данные'
    )
```

the only little issue is that this straightforward approach leads to mypy error:
```
Argument 1 to "as_mutable" of "Mutable" has incompatible type "type[JSONB]"; expected "TypeEngine[Any]" 
```

**Reply by zzzeek** [reply]
instantiate it, `as_mutable(JSONB())`

**Reply by funkindy** [reply]
Thanks! It works fine if i also add `--allow-untyped-calls` flag for mypy run. Otherwise i get `error: Call to untyped function "JSONB" in typed context `

**Reply by CaselIT** [reply]
that type appear not be be typed yet. We do accept PR to add types, if you have the time
